### PR TITLE
[DO NOT MERGE] Diff the 2022-02-01 and 2023-03-31 Rooms API

### PR DESCRIPTION
**Pull request by minnieliu**
# Data Plane API - Pull Request

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*engagement record*](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/271/Azure-REST-API-Stewardship?anchor=rest-api-stewardship-process).

* Link to engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [ ] a public preview
- [ ] GA release

### Change Scope
<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous Open API document (swagger) if applicable, and the root paths that have been updated. </sup>
* Design Document:
* Previous Open API Doc:
* Updated paths:

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azapi) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azapi).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

### CODE REVIEW

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json [hunk 1]
```diff
@@ -301,19 +299,19 @@
           },
           {
             "in": "body",
-            "name": "addParticipantsRequest",
-            "description": "Participants to be added to the room.",
+            "name": "patchParticipantsRequest",
+            "description": "Participants to be patched to the room.",
```
Comment (petrsvihlik) [inline specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:303]:
```suggestion
            "description": "An updated set of participants of the room.",
```

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json [hunk 2]
```diff
@@ -495,69 +371,65 @@
...
+    "InvitedRoomParticipant": {
+      "description": "A participant invited to the room.",
+      "required": [
+        "rawId"
+      ],
       "type": "object",
       "properties": {
         "rawId": {
-          "description": "Raw id of the identifier. Optional in requests, required in responses.",
+          "description": "Raw id of the identifier.",
```
Comment (petrsvihlik) [inline specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:396]:
```suggestion
          "description": "Raw ID representation of the communication identifier.",
```

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json [hunk 3]
```diff
@@ -292,7 +290,7 @@
           {
             "in": "path",
             "name": "roomId",
-            "description": "Room id to add participants.",
+            "description": "Room id to patch participants.",
```
Comment (petrsvihlik) [inline specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:293]:
```suggestion
            "description": "The identifier of the room to update the participants in.",
```

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json [hunk 4]
```diff
@@ -335,158 +333,36 @@
...
-        }
-      }
     }
   },
   "definitions": {
     "CommunicationErrorResponse": {
       "$ref": "../../../Common/stable/2021-03-07/common.json#/definitions/CommunicationErrorResponse"
     },
+    "CommunicationError": {
+      "$ref": "../../../Common/stable/2021-03-07/common.json#/definitions/CommunicationError"
```
Comment (petrsvihlik) [inline specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:343]: pls use the newer version of the common swagger - https://github.com/Azure/azure-rest-api-specs/tree/main/specification/communication/data-plane/Common/stable/2022-07-13

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json [hunk 5]
```diff
@@ -495,69 +371,65 @@
...
-        "communicationIdentifier"
+        "rawId",
+        "role"
       ],
       "type": "object",
       "properties": {
-        "communicationIdentifier": {
-          "$ref": "#/definitions/CommunicationIdentifierModel"
+        "rawId": {
+          "description": "Raw id of the identifier.",
```
Comment (petrsvihlik) [inline specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:380]:
```suggestion
          "description": "Raw ID representation of the communication identifier.",
```

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json [hunk 6]
```diff
@@ -495,69 +371,65 @@
     "RoomParticipant": {
       "description": "A participant of the room.",
       "required": [
-        "communicationIdentifier"
+        "rawId",
+        "role"
       ],
       "type": "object",
       "properties": {
-        "communicationIdentifier": {
```
Comment (petrsvihlik) [inline specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:502]: What's the reason for turning away from the `communicationIdentifier` (that's used by chat, callautomation, mediacomposition)?
When calling `GET /rooms/{roomId}/participants` from an SDK, how are you planning to provide strongly-typed `CommunicationIdentifier` models back to Contoso? 
You could possibly achieve that by [`CommunicationIdentifier.FromRawId(rawId);`](https://learn.microsoft.com/en-us/azure/communication-services/concepts/identifiers?pivots=programming-language-csharp) but it'd be good to agree on the approach and the strategy for maintaining the `CommunicationIdentifierSerializer`.

**Review by petrsvihlik — CHANGES_REQUESTED** [review]

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json (hunk 6 above)
Comment (DominikMe) [reply specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:502]: @petrsvihlik I had advised to embrace raw ids in the service but then transform them to strongly-typed identifiers in the SDK. Getting rid of identifiers in the service allows for a much cleaner RESTful service without RPC-like POST actions and instead embracing PATCH with json+merge+patch. In the future raw id can also be used as a participant id in the /participants collection for single-item APIs.

I'm confident that we can come up with a good transformation in the SDK building on top of our new Identifier<-> RawId conversion APIs. Thoughts?

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json (hunk 5 above)
Comment (DominikMe) [reply specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:380]: That aligns with the doc https://learn.microsoft.com/en-us/azure/communication-services/concepts/identifiers?pivots=programming-language-rest#raw-id-representation, so fine by me :-)
Maybe we could even link the article here?

Comment (mikehang-msft) [reply specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:380]: So will keep this description as we are planning not to use communication identifier.

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json (hunk 4 above)
Comment (mikehang-msft) [reply specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:343]: Updated

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json (hunk 1 above)
Comment (mikehang-msft) [reply specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:303]: Updated

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json (hunk 3 above)
Comment (mikehang-msft) [reply specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:293]: Updated

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json (hunk 2 above)
Comment (mikehang-msft) [reply specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:396]: Similar to above comment, keep the original description

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json (hunk 6 above)
Comment (petrsvihlik) [reply specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:502]: @DominikMe agreed. Just wanted to make sure we're all on the same page. ;)

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json (hunk 5 above)
Comment (petrsvihlik) [reply specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:380]: I wasn't necessarily referring to the `CommunicationIdentifierModel` from the Common swagger. I just wanted to be more explicit about what kind of identifier it is. (Even though you're not depending on the `CommunicationIdentifierModel` here, the rawId still a form of a "communication identifier" and you're still gonna use `CommunicationIdentifier` and derived classes to represent it in the SDK.)

Linking the docs above would help too. I like that idea.

Comment (mikehang-msft) [reply specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json]: @petrsvihlik, thanks for the context. I was under impression its the same reason as not using CommunicationIdentiferModel in this spec. I have updated and added the doc link (sample below). Can you please take a 2nd look?

"raw id of the identifier. Please refer to the following document for additional information on Raw id.
https://learn.microsoft.com/en-us/azure/communication-services/concepts/identifiers?pivots=programming-language-rest#raw-id-representation"

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json (hunk 2 above)
Comment (mikehang-msft) [reply specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json]: Updated the description and added doc link

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json [hunk 7]
```diff
@@ -495,69 +371,65 @@
...
-          "$ref": "#/definitions/CommunicationIdentifierModel"
+        "rawId": {
+          "description": "Raw id of the identifier. Please refer to the following document for additional information on Raw id. <br> https://learn.microsoft.com/en-us/azure/communication-services/concepts/identifiers?pivots=programming-language-rest#raw-id-representation",
```
Comment (petrsvihlik) [inline specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:380]:
```suggestion
          "description": "Raw id of the identifier. Please refer to the following document for additional information on Raw id. <br> https://learn.microsoft.com/azure/communication-services/concepts/identifiers?pivots=programming-language-rest#raw-id-representation",
```

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json [hunk 8]
```diff
@@ -495,69 +371,65 @@
...
+    "InvitedRoomParticipant": {
+      "description": "A participant invited to the room.",
+      "required": [
+        "rawId"
+      ],
       "type": "object",
       "properties": {
         "rawId": {
-          "description": "Raw id of the identifier. Optional in requests, required in responses.",
+          "description": "Raw id of the identifier. Please refer to the following document for additional information on Raw id. <br> https://learn.microsoft.com/en-us/azure/communication-services/concepts/identifiers?pivots=programming-language-rest#raw-id-representation",
```
Comment (petrsvihlik) [inline specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:396]:
```suggestion
          "description": "Raw id of the identifier. Please refer to the following document for additional information on Raw id. <br> https://learn.microsoft.com/azure/communication-services/concepts/identifiers?pivots=programming-language-rest#raw-id-representation",
```

**Review by petrsvihlik — CHANGES_REQUESTED** [review]
always link docs without culture codes pls :)

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json [hunk 9]
```diff
@@ -273,17 +273,15 @@
...
-      "post": {
+      },
+      "patch": {
         "tags": [
           "Participants"
         ],
-        "summary": "Adds participants to a room. If participants already exist, no change occurs.",
-        "operationId": "Rooms_AddParticipants",
+        "summary": "Upsert participants to a room",
+        "operationId": "Rooms_PatchParticipants",
```
Comment (mbakalov) [inline specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:282]: nit: do you want to call this Rooms_UpdateParticipants (to align with Rooms_UpdateRoom), and also rename other references to "patch" (e.g. "patchParticipantsRequest")? I don't feel too strongly either way, but I see we use the "update" language in e.g. [phonenumbers](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/communication/data-plane/PhoneNumbers/stable/2021-03-07/phonenumbers.json).

Comment (mikehang-msft) [reply specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:282]: Hi @mbakalov thanks for your comments.

For Rooms, we have both PUT and PATCH operations so the PATCH for rooms is used for updates. Hence, naming Rooms models with a term "update". That being said, I see "patch" is being used in Rooms too (e.g. "patchRoomRequest"). This nit seems like inconsistent in Rooms that we may need to make change one way or the other. 

On Participants, there is no PUT but PATCH only, hence naming Participant models with a term "patch". We are using "patch" term consistently across participant models.

Are you ok with keeping the name unchanged for participants?

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json (hunk 7 above)
Comment (mikehang-msft) [reply specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json]: Thanks @petrsvihlik. Updated with languages neutral, learning something new 💯

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json (hunk 8 above)
Comment (mikehang-msft) [reply specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json]: Thanks and updated

File: specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json (hunk 9 above)
Comment (minnieliu) [reply specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:282]: Does "Upsert" make more sense for Participants in this case? This is for updating and insert (also deleting) from the participants list.

Comment (minnieliu) [reply specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:282]: Noted and will ask ARB for advice here. For the time being, keeping as PATCH

**Review by petrsvihlik — APPROVED** [review]

**Pull request closed without merging**